hudi-common/src/main/java/org/apache/hudi/io/storage/HoodieParquetKeyedLookupReader.java

```diff
@@ -0,0 +1,318 @@
+ *
+ * as well as being sorted by `key` column.
+ *
+ * Known limitations:
+ * 1) Does not do bloom filter based skipping of row groups.
+ */
+public class HoodieParquetKeyedLookupReader {
+  private static Logger LOG = LoggerFactory.getLogger(HoodieParquetKeyedLookupReader.class);
+  private static String KEY = "key";
+  private static String VALUE = "value";
+  private final InputFile parquetFile;
+  private final Configuration conf;
+  private final ParquetFileMetadataLoader metadataLoader;
+  private final CompressionCodecFactory codecFactory;
+  private final ParquetMetadataConverter converter;
+
+  public HoodieParquetKeyedLookupReader(Configuration conf, InputFile parquetFile) throws Exception {
+    this.conf = conf;
+    this.parquetFile = parquetFile;
+    this.metadataLoader = new ParquetFileMetadataLoader(
+        parquetFile, ParquetFileMetadataLoader.Options.builder().enableLoadBloomFilters().build());
+    this.codecFactory = HadoopCodecs.newFactory(0);
+    this.converter = new ParquetMetadataConverter();
+
+    metadataLoader.load();
+  }
+
+  public Map<String, Option<String>> lookup(SortedSet<String> keys) throws Exception {
+    Map<String, Option<String>> keyToValue = new HashMap<>();
+    try (CompressionConverter.TransParquetFileReader reader = new CompressionConverter.TransParquetFileReader(
+        parquetFile, HadoopReadOptions.builder(conf).build())) {
+      Map<String, String> matchingRecords = getMatchingRecords(reader, new LinkedList<>(keys));
+      for (String key: keys) {
+        if (matchingRecords.containsKey(key)) {
+          keyToValue.put(key, Option.of(matchingRecords.get(key)));
+        } else {
+          keyToValue.put(key, Option.empty());
+        }
+      }
+    }
+    return keyToValue;
+  }
+
+  private Map<String, String> getMatchingRecords(CompressionConverter.TransParquetFileReader reader,
+                                                 Queue<String> keys) throws Exception {
+    Map<String, String> keyToValue = new HashMap<>();
+    for (RowGroup rowGroup : metadataLoader.getRowGroups()) {
+      final int keyColNumber = searchColumn(rowGroup, KEY);
+      if (keyColNumber < 0) {
+        throw new IllegalArgumentException("Cannot find key column in schema.");
+      }
+
+      // Skip row group using bloom filter if possible.
+      if (shouldSkip(rowGroup, keyColNumber, new LinkedList<>(keys))) {
+        continue;
+      }
+
+      ArrayList<Pair<String, Long>> keyPositions = lookupKeyColumnChunk(reader, rowGroup, keyColNumber, keys);
+      // Now fetch the respective values out.
+      final int valueColNumber = searchColumn(rowGroup, VALUE);
+      if (valueColNumber < 0) {
+        throw new IllegalArgumentException("Cannot find value column in schema.");
+      }
+      if (!keyPositions.isEmpty()) {
+        keyToValue.putAll(fetchFromValueColumnChunk(reader, rowGroup, valueColNumber, keyPositions));
+      }
+      LOG.info("Done with rowGroup");
```

Comment:
move to lower logging level like Debug. Fix logging across the board. IDK if we need anything beyond 1-2 lines in INFO.